### PR TITLE
BackendConfiguration Closes #580

### LINKED ISSUE #580: Introduce BackendConfiguration like pluggable_backends.

### PR DESCRIPTION
**Pull request by mcovarr**

### CODE REVIEW

**Comment by mcovarr** [comment]
@geoffjentry and @Horneth

File: README.md [hunk 1]
```diff
@@ -286,12 +286,12 @@ database {
 For backends that support aborting task invocations, Cromwell can be configured to automatically try to abort all currently running calls (and set their status to `Aborted`) when a SIGINT is sent to the Cromwell process.  To turn this feature on, set the configuration option
```
Comment (mcovarr) [inline README.md]: um yeah, there should be a LOT more changes here... 😦

File: engine/src/main/resources/application.conf [hunk 2]
```diff
@@ -78,86 +70,128 @@ call-caching {
...
-  userAuth {
-    // user = ""
-    // secretsFile = ""
-    // dataStoreDir = ""
-  }
+  application-name = "cromwell"
+
+  auths = [
+    {
+      name = "application-default"
```
Comment (geoffjentry) [inline engine/src/main/resources/application.conf]: i'm assuming this came over from the left hand side but why is it `_` in one and `-` in the other?

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala [hunk 3]
```diff
@@ -84,6 +71,12 @@ trait Backend {
 
   def actorSystem: ActorSystem
 
+  def backendConfigEntry: BackendConfigurationEntry
+
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: F'ing intellij

File: engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 4]
```diff
@@ -266,18 +274,14 @@ case class JesBackend(actorSystem: ActorSystem)
...
-
+    * Get a GcsLocalizing from workflow options iff client secrets and refresh token are available as this method
+    * is called only if an auth.json needs to be written.
+    */
+  def refreshTokenAuth(workflow: WorkflowDescriptor): Option[JesAuthInformation] = {
     for {
-      secrets <- extractSecrets(gcsConf.authMode)
-      token <- workflow.workflowOptions.get(GoogleConfiguration.RefreshTokenOptionKey).toOption
-    } yield GcsLocalizing(secrets, token)
+      clientSecrets <- Seq(jesAttributes.gcsFilesystemAuth) collectFirst { case s: ClientSecrets => s }
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: :person_frowning:

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsPath.scala [hunk 5]
```diff
@@ -6,7 +6,7 @@ import scala.util.{Failure, Success, Try}
 /** Represents a Google Cloud Storage path, like gs://bucket/path/to/object.txt
   *
   * @param bucket - should adhere to https://cloud.google.com/storage/docs/bucket-naming?hl=en#requirements
-  * @param objectName
+  * @param objectName - the name of the thing
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/io/gcs/GcsPath.scala]: res ipsa loquitor

**Comment by geoffjentry** [comment]
https://twitter.com/modocache/status/719524552171220992

**Comment by geoffjentry** [comment]
:+1:

File: engine/src/main/resources/application.conf [hunk 6]
```diff
@@ -78,86 +70,128 @@ call-caching {
...
+
+        filesystems = {
+          gcs {
+            // A reference to a potentially different auth for manipulating files via engine functions.
+            auth = "user-via-refresh"
+          }
+        }
+      }
+    }
+  ]
```
Comment (scottfrazer) [inline engine/src/main/resources/application.conf]: This is looking a lot more intuitive than before!

File: engine/src/main/resources/application.conf [hunk 7]
```diff
@@ -78,86 +70,128 @@ call-caching {
...
-  // For instance data localization and delocalization in GoogleCloudStorage will be done on behalf of another entity (typically a user).
-  // This allows cromwell to localize data that wouldn't be accessible using whatever "cromwellAuthenticationScheme" has been defined.
-  // The currently only available scheme is refresh, which requires a refresh token to be passed upon every workflow submission (see "workflow options" in cromwell documentation).
-  // In that scheme, a client_id and client_secrets must be provided. They must be the ones that have been used to generate the above-mentioned refresh token.
-  // userAuthenticationScheme = "refresh"
-  // refreshTokenAuth = {
-  //   client_id = ""
-  //   client_secret = ""
-  // }
+backend {
```
Comment (francares) [inline engine/src/main/resources/application.conf]: Shouldn't this be backend package?

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala [hunk 8]
```diff
@@ -84,6 +71,12 @@ trait Backend {
 
   def actorSystem: ActorSystem
 
+  def backendConfigEntry: BackendConfigurationEntry
```
Comment (francares) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: Is this an intermediate step to then later move this code to backend package?

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala [hunk 9]
```diff
@@ -201,12 +194,12 @@ trait Backend {
   /**
     * Compute a hash that uniquely identifies this call
     */
-  def hash(descriptor: BackendCallJobDescriptor)(implicit ec: ExecutionContext): Future[ExecutionHash] = {
+  def hash(descriptor: BackendCallJobDescriptor, credential: Option[Credential] = None)(implicit ec: ExecutionContext): Future[ExecutionHash] = {
```
Comment (francares) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: will this function go away in the future right?

File: engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala [hunk 10]
```diff
@@ -16,13 +16,19 @@ import scala.util.Try
...
   */
 object CromwellBackend {
-  private val log = LoggerFactory.getLogger(getClass.getName)
   private var _backends: Option[Map[String, Backend]] = None
   private var _defaultBackend: Option[Backend] = None
 
-  def initBackends(backendTypes: List[String], defaultBackend: String, actorSystem: ActorSystem) = {
-    _backends = Option((backendTypes map { backendType => (backendType, Backend.from(backendType, actorSystem)) }).toMap)
-    _defaultBackend = Option(Backend.from(defaultBackend, actorSystem))
+  def backendFromConfig(entry: BackendConfigurationEntry, actorSystem: ActorSystem): Backend = {
```
Comment (francares) [inline engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala]: I think this belongs to Backend package.

File: engine/src/main/scala/cromwell/engine/backend/EnhancedWorkflowOptions.scala [hunk 11]
```diff
@@ -0,0 +1,16 @@
+package cromwell.engine.backend
+
+import cromwell.core.WorkflowOptions
+import cromwell.filesystems.gcs.GoogleAuthMode
+import cromwell.filesystems.gcs.GoogleAuthMode.GoogleAuthOptions
+
+import scala.util.Try
+
+object EnhancedWorkflowOptions {
```
Comment (francares) [inline engine/src/main/scala/cromwell/engine/backend/EnhancedWorkflowOptions.scala]: Is this JES specific?

File: engine/src/main/resources/application.conf (hunk 2 above)
Comment (mcovarr) [inline engine/src/main/resources/application.conf]: hrm.  I tried to be consistent in dashing the keys, the values seem to be kind of a mish mash.  Do you think everything should be dashes or just consistent?

File: engine/src/main/resources/application.conf (hunk 7 above)
Comment (mcovarr) [inline engine/src/main/resources/application.conf]: No, this specifies which backends are configured in this Cromwell instance, as well as how they are configured.

File: engine/src/main/resources/application.conf (hunk 6 above)
Comment (mcovarr) [inline engine/src/main/resources/application.conf]: 😄

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala (hunk 8 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: I wasn't really addressing that point with these changes, but I certainly wouldn't argue that this class belongs in the backend project.  😄

File: engine/src/main/resources/application.conf (hunk 2 above)
Comment (geoffjentry) [inline engine/src/main/resources/application.conf]: Consistent would be nice but it's not like we're perfect about this everywhere anyways. I think the last time we tried to hash out the - vs _ thing I made strong claims for one which was exactly the opposite of what I'd said 2 times ago so I'll refrain from saying anything this time :)

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala (hunk 9 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: That's not a concern of these changes

File: engine/src/main/scala/cromwell/engine/backend/EnhancedWorkflowOptions.scala (hunk 11 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/EnhancedWorkflowOptions.scala]: JES **requires** this, but our other two backends can use it as well.

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala (hunk 8 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: There are still some engine dependencies here, but it looks like we're getting closer to being able to move this.

File: engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala (hunk 10 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala]: I wouldn't dispute that, and here it looks like there aren't engine dependencies.

File: engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala (hunk 4 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: I'll grant that is weird but is there something in particular that would make you 😺 ?

Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Oh I was just lamenting the `Seq`, feel free to ignore me

Aside: at some point in the last few weeks gmail has started using some bizarre emojis for github stuff

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsPath.scala (hunk 5 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/io/gcs/GcsPath.scala]: yeah, can't win.  The first bit of doc adds real value, but if I don't put **something** here IntelliJ whines.

Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/io/gcs/GcsPath.scala]: F'ing intellij

File: engine/src/main/scala/cromwell/engine/backend/Backend.scala (hunk 8 above)
Comment (francares) [inline engine/src/main/scala/cromwell/engine/backend/Backend.scala]: Oh, OK... np...

File: engine/src/main/resources/application.conf [hunk 12]
```diff
@@ -78,86 +70,128 @@ call-caching {
...
+  default = "Local"
+  providers = [
+    {
+      name = "Local"
+      class = "cromwell.engine.backend.local.LocalBackend"
```
Comment (cjllanwarne) [inline engine/src/main/resources/application.conf]: I think we'll eventually want to name this differently - if we want to allow the optional initializationClass, executionClass, finalizationClass, ...

File: engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala (hunk 4 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: That's an interesting interpretation of "person frowning". I wouldn't have guessed

File: engine/src/main/resources/application.conf [hunk 13]
```diff
@@ -78,86 +70,128 @@ call-caching {
...
+      name = "application-default"
+      scheme = "application_default"
+    },
+    {
+      name = "user-via-refresh"
+      scheme = "refresh_token"
+      client-id = "secret_id"
+      client-secret = "secret_secret"
+    }
+  ]
```
Comment (Horneth) [inline engine/src/main/resources/application.conf]: Should we also add an example for service account ?

File: engine/src/main/resources/application.conf [hunk 14]
```diff
@@ -78,86 +70,128 @@ call-caching {
...
+      class = "cromwell.engine.backend.jes.JesBackend"
+      config {
+        // Google project
+        project = "my-cromwell-workflows"
+
+        // Base bucket for workflow executions
+        root = "gs://my-cromwell-workflows-bucket"
+
+        // Endpoint for APIs, no reason to change this unless directed by Google.
+        endpoint-url = "https://genomics.googleapis.com/"
```
Comment (Horneth) [inline engine/src/main/resources/application.conf]: Could we move this to the `genomics` stanza ?

File: engine/src/main/scala/cromwell/engine/backend/CanUseGcsFilesystem.scala [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+  protected def gcsFilesystem(options: WorkflowOptions): Option[FileSystem] = {
+
+    def gcsFilesystemMustHaveAuth: Option[String] = {
+      backendConfig.getConfigOption("filesystems.gcs") map {
+        _.getStringOr("auth", throw new RuntimeException(s"No auth in filesystems.gcs for backend '$name'.")) }
+    }
+
+    def gcsFilesystemAuthMustBeLegit(authName: String): GoogleAuthMode = {
+      def prependName(errs: NonEmptyList[String]): String = s"For backend '$name': " + errs.list.mkString("\n")
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/backend/CanUseGcsFilesystem.scala]: Is it possible to use https://github.com/broadinstitute/lenthall/blob/develop/src/main/scala/lenthall/exception/ExceptionAggregation.scala#L7 ?
It preserves the list of errors separate form the message so we can report / display them more easily

File: engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala [hunk 16]
```diff
@@ -16,13 +16,19 @@ import scala.util.Try
...
+    val constructor = Class.forName(entry.className).getConstructor(classOf[BackendConfigurationEntry], classOf[ActorSystem])
+    constructor.newInstance(entry, actorSystem).asInstanceOf[Backend]
+  }
+
+  def initBackends(backendEntries: List[BackendConfigurationEntry], defaultBackendEntry: BackendConfigurationEntry, actorSystem: ActorSystem) = {
+    case class EntryAndBackend(entry: BackendConfigurationEntry, backend: Backend)
+    val entriesAndBackends = backendEntries map { e => EntryAndBackend(e, backendFromConfig(e, actorSystem)) }
+    _defaultBackend = entriesAndBackends collectFirst { case eb if eb.entry == defaultBackendEntry => eb.backend }
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala]: `find` ?

File: engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala [hunk 17]
```diff
@@ -16,13 +16,19 @@ import scala.util.Try
...
-  private val log = LoggerFactory.getLogger(getClass.getName)
   private var _backends: Option[Map[String, Backend]] = None
   private var _defaultBackend: Option[Backend] = None
 
-  def initBackends(backendTypes: List[String], defaultBackend: String, actorSystem: ActorSystem) = {
-    _backends = Option((backendTypes map { backendType => (backendType, Backend.from(backendType, actorSystem)) }).toMap)
-    _defaultBackend = Option(Backend.from(defaultBackend, actorSystem))
+  def backendFromConfig(entry: BackendConfigurationEntry, actorSystem: ActorSystem): Backend = {
+    val constructor = Class.forName(entry.className).getConstructor(classOf[BackendConfigurationEntry], classOf[ActorSystem])
+    constructor.newInstance(entry, actorSystem).asInstanceOf[Backend]
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala]: Thinking about this, am I right in saying that the Backend interface is a trait and only defines things that needs to be implemented, nowhere do we actually enforce what the backend constructor has to look like. Could that be a problem to generically instantiate backends ?

File: engine/src/main/resources/application.conf (hunk 13 above)
Comment (mcovarr) [inline engine/src/main/resources/application.conf]: sure

File: engine/src/main/resources/application.conf (hunk 12 above)
Comment (mcovarr) [inline engine/src/main/resources/application.conf]: yes absolutely, this is just to fit our current reality.

File: engine/src/main/resources/application.conf (hunk 14 above)
Comment (mcovarr) [inline engine/src/main/resources/application.conf]: Having it here allows for different JES backend instances which can point to production or staging within the same Cromwell.

File: engine/src/main/scala/cromwell/engine/backend/CanUseGcsFilesystem.scala (hunk 15 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/CanUseGcsFilesystem.scala]: Lenthall doesn't seem to already have something that exactly fits this, but yeah:

```scala
class ErrorMessageAggregatedException(override val exceptionContext: String,
                                      override val errorMessages: Traversable[String]) extends MessageAggregation

GoogleConfiguration.Instance.auth(authName) valueOr { errs => throw new ErrorMessageAggregatedException(s"For backend '$name': ", errs.list) }
```

File: engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala (hunk 17 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala]: Yes I think that's right, there has to be some contract for how to instantiate a backend.

File: engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala (hunk 16 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala]: There's a mapping happening there too.  This is equivalent:

```scala
_defaultBackend = entriesAndBackends find { _.entry == defaultBackendEntry } map { _.backend }
```

File: engine/src/main/resources/application.conf (hunk 14 above)
Comment (mcovarr) [inline engine/src/main/resources/application.conf]: oh I first read this as `google` and not `genomics`; yeah, that makes sense.

File: README.md (hunk 1 above)
Comment (scottfrazer) [inline README.md]: The table of contents also needs to be regenerated, I'll do that.

**Comment by Horneth** [comment]
👍

**Pull request merged**